File: solution/156BinaryTreeUpsideDown/156BinaryTreeUpsideDown.cpp
#include <sstream>
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <algorithm>
#include <limits>

using namespace std;

/*
Given a binary tree where all the right nodes are either leaf nodes with a sibling 
(a left node that shares the same parent node) or empty, flip it upside down and turn
it into a tree where the original right nodes turned into left leaf nodes. Return the new root.

For example:
Given a binary tree {1,2,3,4,5},
    1
   / \
  2   3
 / \
4   5
return the root of the binary tree [4,5,2,#,#,3,1].
   4
  / \
 5   2
    / \
   3   1  
confused what "{1,#,2,3}" means? > read more on how binary tree is serialized on OJ.
*/

//Definition for binary tree
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode *upsideDownBinaryTree(TreeNode *root) {
    	if(root==NULL) return NULL;
    	
    	TreeNode *curr_left, *curr_right;
    	TreeNode *curr, *parent, *slibe;

    	parent = root;
    	slibe = root->right;
    	curr = root->left;
    	
    	root->left = root->right = NULL;

    	while(curr!=NULL){
	    	curr_left = curr->left;
	    	curr_right= curr->right;

	    	curr->left = slibe;
	    	curr->right= parent;

	    	parent = curr;
	    	slibe  = curr_right;
	    	curr   = curr_left;
	    }

	    return parent;
    }
};

int main(){
	Solution s;

	return 0;
} 

File: src/156BinaryTreeUpsideDown/156BinaryTreeUpsideDown.go
package main

import "fmt"

/**
 * Definition for a binary tree node.
 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func upsideDownBinaryTreeHelper(root *TreeNode) (*TreeNode, *TreeNode) {
	if root.Left == nil {
		return root, root
	}
	newRoot, right := upsideDownBinaryTreeHelper(root.Left)
	right.Left = root.Right
	right.Right = root
	root.Left = nil
	root.Right = nil
	return newRoot, root
}
func upsideDownBinaryTree(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	root, _ = upsideDownBinaryTreeHelper(root)
	return root
}

func main() {
	n4 := &TreeNode{4, nil, nil}
	n5 := &TreeNode{5, nil, nil}
	n2 := &TreeNode{2, n4, n5}
	n3 := &TreeNode{3, nil, nil}
	n1 := &TreeNode{1, n2, n3}
	root := upsideDownBinaryTree(n1)
	fmt.Printf("\n")
	for root != nil {
		fmt.Printf("%d ", root.Val)
		if root.Left != nil {
			fmt.Printf("%d ", root.Left.Val)
		}
		if root.Right != nil {
			//fmt.Printf("%d ", root.Right.Val)
		}
		root = root.Right
	}
}
